### DISCUSSION TITLE
Displaying images in Emacs in Kitty

### DISCUSSION THREAD

**Discussion by gzagatti** [kovidgoyal/kitty#4664] [Q&A] [answered]
I was wondering if anyone had any luck displaying images inside of Emacs inside of kitty. Is that even possible? If so, how can we do it?

I wanted to display inline images in `org` files and `ein` notebooks. Thus, it would be very useful for my workflow.

The function [`display-images-p`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Images.html) suggests that this is not possible in kitty. But I'm wondering if there's a way around it.

**Comment by page-down** [comment, marked as answer]
It looks like this requires the solution discussed in the following link.

https://github.com/kovidgoyal/kitty/discussions/4021

**Reply by peterjc** [reply]
Did this functionality get added to emacs? https://www.gnu.org/software/emacs/manual/html_node/elisp/Display-Property.html doesn't seem to talk about how the images are actually handled (and if this is in X-Windows emacs only, or terminal emacs too).

**Comment by kovidgoyal** [comment]
Far better would be for someone to add support for the full kitty graphics protocol to emacs. https://sw.kovidgoyal.net/kitty/graphics-protocol/

**Reply by gzagatti** [reply]
Thanks for pointing it out! I'm glad it's possible. Unfortunately, it's well beyond my capability :disappointed:

**Comment by gzagatti** [comment]
I sort of got what I wanted by writing a custom kitten that will create a new window and then cat the image that I want to see from Emacs. In Emacs, I only need to set a custom `mailcap-user-mime-data` such that it automatically uses my kitten when opening an image in an external viewer.

The kitten looks like below. I'm not sure if that's the best way to send a command to the other window. For instance, nothing happens if the pointer in the other window is in normal mode. How could I make sure that the command simply gets sent to the shell and executed?

Also, `boss.launch` does not return a window, so I have to search for it just after creating it. However, [the source code](https://github.com/kovidgoyal/kitty/blob/696f371aa4478902b79daf1d24c4183283d3e20c/kitty/launch.py#L452) seems to suggest that a window is returned.

```python
# imgviewer.py
from typing import List
from pathlib import Path
from kitty.boss import Boss
from kittens.tui.handler import result_handler

def main(args: List[str]) -> str:
    pass

@result_handler(no_ui=True)
def handle_result(args: List[str], result: None, target_window_id: int, boss: Boss) -> None:
    args = args[1:]
    files = []
    w = boss.window_id_map.get(target_window_id)
    for arg in args:
        file = Path(arg).expanduser()
        if file.exists():
            files.append(file)
    if len(files) == 0:
        return
    try:
        imgviewer = next(boss.match_windows(f"title:imgviewer{w.id}"))
    except StopIteration:
        imgviewer = boss.launch("--window-title", f"imgviewer{w.id}",
            "--cwd", "current", "--keep-focus")
        imgviewer = next(boss.match_windows(f"title:imgviewer{w.id}"))
        imgviewer.write_to_child("export HISTFILE=\r")
    imgviewer.write_to_child(f"timg {' '.join(str(f) for f in files)}\r")
```

And in Emacs, I add the following to the configuration file:

```elisp
(custom-set-variables                                                                                                     
	'(mailcap-user-mime-data '(("kitty @ kitten ./imgviewer.py %s" "image/png" nil))))
```

This is how it looks like:

![image](https://user-images.githubusercontent.com/11043156/153338926-a7c48c89-12ee-43bd-a7df-e46e309bb931.png)

**Comment by kovidgoyal** [comment]
You dont need to do any of that if you just want to open the image in a
new window. You can just do

(mailcap-user-mime-data '(("kitty kitty +kitten icat --hold %s"))'

**Reply by gzagatti** [reply]
Thanks. I wanted to persist the window such that it becomes sort of a dump for any image I want to cat when during my session. This is the reason why I ended up developing the proof-of-concept kitten. It was also a good chance to learn how to create one :smile:

**Comment by kovidgoyal** [comment]
Ah cool. Note that this can be done purely by using the remote control
api without needing a kitten, but of course a kitten works as well.

**Comment by retpolanne** [comment]
Chiming in with a huge thread about implementing kitty protocols on emacs

https://lists.nongnu.org/archive/html/emacs-devel/2022-09/msg00384.html

**Comment by vonpupp** [comment]
I am a new user of Kitty. Be able to see inline images of emacs within Kitty would be just amazing!

**Comment by lf-araujo** [comment]
Once the mimecap custom variable is added, how to open the link? Clicking a link in orgmode still opens the binary within emacs.

**Comment by afermg** [comment]
If anyone visits this thread after March 2026, a proof of concept was developed (via Claude) [github.com/cashmeredev/kitty-graphics.el](https://github.com/cashmeredev/kitty-graphics.el). It is a bit bare-bones, but it works for me on both Kitty and Wezterm.
